File: backend/app/utils/trip_permissions.py
from typing import Optional
from bson import ObjectId
from fastapi import HTTPException, status
from app.database import get_database


async def check_trip_access(trip_id: str, user_id: str, require_role: Optional[str] = None) -> dict:
    """Check if user has access to trip, optionally requiring a specific role"""
    db = get_database()
    trip = await db.trips.find_one({"_id": ObjectId(trip_id)})
    
    if not trip:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Trip not found")
    
    user_obj_id = ObjectId(user_id)
    member = next(
        (m for m in trip.get("members", []) if m.get("userId") == user_obj_id),
        None
    )
    
    if not member:
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Access denied")
    
    if require_role and member.get("role") != require_role:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail=f"Requires {require_role} role"
        )
    
    return trip

File: backend/src/utils/trip_permissions.ts
import { ObjectId } from 'mongodb';
import { getDatabase } from '../database.js';

export interface TripDocument {
  _id: ObjectId;
  members: Array<{ userId: ObjectId; role: string }>;
  [key: string]: any;
}

export async function checkTripAccess(
  tripId: string,
  userId: string,
  requireRole?: 'owner' | 'editor' | 'viewer'
): Promise<TripDocument> {
  const db = getDatabase();
  const trip = await db.collection('trips').findOne({ _id: new ObjectId(tripId) });

  if (!trip) {
    throw new Error('Trip not found');
  }

  const userObjId = new ObjectId(userId);
  const member = trip.members?.find((m: { userId: ObjectId }) => m.userId.equals(userObjId));

  if (!member) {
    throw new Error('Access denied');
  }

  if (requireRole && member.role !== requireRole) {
    throw new Error(`Requires ${requireRole} role`);
  }

  return trip as TripDocument;
}
